causalpy/experiments/base.py

```diff
@@ -0,0 +1,63 @@
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Base class for quasi experimental designs.
+"""
+
+from abc import abstractmethod
+
+from causalpy.pymc_models import PyMCModel
+from causalpy.skl_models import ScikitLearnModel
+
+
+class BaseExperiment:
+    """Base class for quasi experimental designs."""
```

Comment:
It might be either as the reader to have `supports_ols` `supports_bayes` class attributes. Just because those are at the top of the child classes. Or use type hints for the models instead of just None

Otherwise I have to get into the weeds of the code for each.

First might be like this
```python
class BaseExperiment:
    """Base class for quasi experimental designs."""
    supports_bayes: bool
    supports_ols: bool
   
    def __init__(self, model) -> None:
        if model is not None:
            self.model = model

        if isinstance(self.model, PyMCModel) and not self.supports_bayes: raise ...
        if isinstance(self.model, ScikitLearnModel) and not self.supports_ols: raise ...

        if self.model is None:
            raise ValueError("fitting_model not set or passed.")
    ...


class DifferenceInDifference(BaseExperiment): 
    supports_ols = True
    supports_bayes = True


class RegressionKink(BaseExperiment): 
    supports_bayes = True
    supports_ols = False
```

Reply:
This is a great idea. I've implemented this in https://github.com/pymc-labs/CausalPy/pull/381/commits/cc62438e7adbc30e68fbcd25044475d57e7aa758. I've also added tests to check that exceptions are raised when providing old models to experiment classes that only support bayesian models. These can be removed over time if that support is added.